### PR TITLE
Pact testing

### PR DESCRIPTION
**Pull request by mhisltd**
**Before creating a pull request make sure that:**
This is just a dry run to see if the pipeline publishes a pact to broker.
- [ ] commit messages are meaningful and follow good commit message guidelines
- [ ] README and other documentation has been updated / added (if needed)
- [ ] tests have been updated / new tests has been added (if needed)
- [ ] the `keep-helm` label has been added, if the helm release should be persisted after a successful build

Please remove this line and everything above and fill the following sections:

### Change description ###

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ ] No
```

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -147,6 +172,31 @@ pmd {
     ruleSetFiles = files("config/pmd/ruleset.xml")
 }
 
+runAndPublishConsumerPactTests.dependsOn contract
+
+runAndPublishConsumerPactTests.finalizedBy pactPublish
+
+project.ext {
+  pacticipantVersion = getCheckedOutGitCommitHash()
+}
```
Comment (rapolaskaseliscgi) [inline build.gradle:181]: I believe, custom property can be added in this manner:
`project.ext.participantVersion = getCheckedOutGitCommitHash()`
And if that works, I think it is nicer.

File: build.gradle [hunk 2]
```diff
@@ -6,14 +6,15 @@ plugins {
...
     id 'io.spring.dependency-management' version '1.1.0'
-    id 'net.serenity-bdd.serenity-gradle-plugin' version '3.7.0'
-    id 'org.flywaydb.flyway' version '9.20.0'
+    id 'net.serenity-bdd.serenity-gradle-plugin' version '3.9.0'
+    id 'org.flywaydb.flyway' version '9.21.0'
     id 'org.owasp.dependencycheck' version '8.3.1'
-    id 'org.sonarqube' version '4.2.1.3168'
-    id 'org.springframework.boot' version '3.1.1'
+    id 'org.sonarqube' version '4.3.0.3225'
+    id 'org.springframework.boot' version '3.1.2'
```
Comment (rapolaskaseliscgi) [inline build.gradle:13]: I'd suggest against doing these lib updates and better do a rebase on master that should bring your branch up to date (though that rebase might be a bit painful)

File: build.gradle [hunk 3]
```diff
@@ -382,6 +433,13 @@ dependencies {
     testCompileOnly group: 'org.projectlombok', name: 'lombok'
     testAnnotationProcessor group: 'org.projectlombok', name: 'lombok'
 
+    //pact contract testing
+    contractTestImplementation group: 'au.com.dius.pact.consumer', name: 'junit5', version: '4.4.0-beta.2'
+    contractTestImplementation group: 'org.springframework.boot', name: 'spring-boot-starter-test', version: '2.7.12'
```
Comment (rapolaskaseliscgi) [inline build.gradle:438]: are you sure we need spring boot version 2.7? couple months ago we have migrated to spring boot v3.

File: src/contractTest/java/uk/gov/hmcts/reform/laubackend/idam/pacttests/IdamApiConsumerTest.java [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+import java.util.Map;
+
+import static org.junit.Assert.assertNotNull;
+import static org.springframework.http.MediaType.APPLICATION_FORM_URLENCODED_VALUE;
+
+@SuppressWarnings({"unchecked", "PMD.AvoidDuplicateLiterals"})
+public class IdamApiConsumerTest extends IdamConsumerTestBase {
+
+    private final String lauIdamBackEndApiUrl = "";
```
Comment (rapolaskaseliscgi) [inline src/contractTest/java/uk/gov/hmcts/reform/laubackend/idam/pacttests/IdamApiConsumerTest.java:24]: Is it used anywhere?

File: src/contractTest/java/uk/gov/hmcts/reform/laubackend/idam/pacttests/IdamConsumerTestBase.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+@ExtendWith(PactConsumerTestExt.class)
+@TestInstance(TestInstance.Lifecycle.PER_CLASS)
+@PactTestFor(providerName = "idamApi_oidc", port = "8891")
+@PactDirectory("pacts")
+@SpringBootTest({
+    "idam.api.url : localhost:8891"
+})
+@TestPropertySource(locations = {"/config/application_contract.properties"})
+public abstract class IdamConsumerTestBase {
+    public static final int SLEEP_TIME = 2000;
```
Comment (rapolaskaseliscgi) [inline src/contractTest/java/uk/gov/hmcts/reform/laubackend/idam/pacttests/IdamConsumerTestBase.java:26]: I would prefer to see units as well, e.g. `SLEEP_TIME_IN_MS`.

File: src/functionalTest/java/uk/gov/hmcts/reform/laubackend/idam/serenityfunctionaltests/config/EnvConfig.java
Comment (rapolaskaseliscgi) [inline src/functionalTest/java/uk/gov/hmcts/reform/laubackend/idam/serenityfunctionaltests/config/EnvConfig.java:1]: This change shouldn't be here... the same with below. Those changes already on master branch

**Review by rapolaskaseliscgi — COMMENTED** [review]
Please, rebase on master branch, preferably squashing all the commits. That action should make PR smaller and to the point

File: src/contractTest/java/uk/gov/hmcts/reform/laubackend/idam/pacttests/IdamApiConsumerTest.java (hunk 4 above)
Comment (mhisltd) [reply src/contractTest/java/uk/gov/hmcts/reform/laubackend/idam/pacttests/IdamApiConsumerTest.java:24]: Removed

File: src/contractTest/java/uk/gov/hmcts/reform/laubackend/idam/pacttests/IdamConsumerTestBase.java (hunk 5 above)
Comment (mhisltd) [reply src/contractTest/java/uk/gov/hmcts/reform/laubackend/idam/pacttests/IdamConsumerTestBase.java:26]: Done

File: build.gradle (hunk 1 above)
Comment (mhisltd) [reply build.gradle]: Done

File: build.gradle (hunk 3 above)
Comment (mhisltd) [reply build.gradle]: Removed this line as we dont need it

File: src/functionalTest/java/uk/gov/hmcts/reform/laubackend/idam/serenityfunctionaltests/config/EnvConfig.java
Comment (mhisltd) [reply src/functionalTest/java/uk/gov/hmcts/reform/laubackend/idam/serenityfunctionaltests/config/EnvConfig.java:1]: This is removed from the change after rebase

**Review by rapolaskaseliscgi — APPROVED** [review]

File: src/contractTest/resources/config/application_contract.properties [hunk 6]
```diff
@@ -0,0 +1,8 @@
+core_case_data.api.url=${ccd_casedatastore_baseurl:http://localhost:3451}
+ccd.jurisdictionid=SSCS
+ccd.casetype=Benefit
+ccd.eventid.create=createTestCase
+idam.api.url=${IDAM_API_URL:http://localhost:8891}
+idam.oauth2.user.email=${IDAM_LAU_SYSTEMUPDATE_USER:system.update@hmcts.net}
+idam.oauth2.user.password=${IDAM_LAU_SYSTEMUPDATE_PASSWORD:Pa55word11}
+idam.client.secret=${IDAM_OAUTH2_CLIENT_SECRET:QM5RQQ53LZFOSIXJ}
```
[the remark this answers is not in the document]
Comment (mhisltd) [reply src/contractTest/resources/config/application_contract.properties:8]: This is taken from SSCS implementation so should not be sensitive.

**Pull request merged**